// @react-native-community/cli/build/install/uninstall.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.default = void 0;

function _child_process() {
  const data = require('child_process');

  _child_process = function() {
    return data;
  };

  return data;
}

var _logger = _interopRequireDefault(require('../util/logger'));

var _PackageManager = _interopRequireDefault(require('../util/PackageManager'));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */
const spawnOpts = {
  stdio: 'inherit',
  stdin: 'inherit'
};

async function uninstall(args, ctx) {
  const name = args[0];
  const res = (0, _child_process().spawnSync)(
    'react-native',
    ['unlink', name],
    spawnOpts
  );

  if (res.status) {
    process.exit(res.status);
  }

  new _PackageManager.default({
    projectDir: ctx.root
  }).uninstall([name]);

  _logger.default.info(
    `Module ${name} has been successfully uninstalled & unlinked`
  );
}

var _default = {
  func: uninstall,
  description: 'uninstall and unlink native dependencies',
  name: 'uninstall <packageName>'
};
exports.default = _default;
